### PR TITLE
[16.0][IMP] delivery_driver: Flag driver in partner

### PR DESCRIPTION
**Pull request by EmilioPascual**
- When a driver is assigned in a shipping method, the partner is marked as a driver.
- Unmarking partners as drivers is not allowed if it is associated with a sending method.
- In pickings and move lines add domain in driver. Only partners marked as driver.
- When selecting a partner driver, only the name of the partner is displayed.
- If driver is changed in picking, also it's changed in origin picking (if you activated multi-steps routes)

MT-5232 @moduon

@rafaelbn @yajo @shide @gelojr please review.

### CODE REVIEW

File: delivery_driver/models/stock_picking.py [hunk 1]
```diff
@@ -11,14 +11,19 @@ class StockPicking(models.Model):
...
         readonly=False,
     )
 
-    @api.depends("carrier_id")
+    @api.depends("carrier_id", "move_ids.move_dest_ids.picking_id.driver_id")
     def _compute_driver_id(self):
         for picking in self:
             if picking.state not in {"done", "cancel"}:
-                picking.driver_id = picking.carrier_id.driver_id
+                driver = picking.move_ids.move_dest_ids.mapped("picking_id.driver_id")
```
Comment (Shide) [inline delivery_driver/models/stock_picking.py]: mapped should contain move_dest_ids?

```suggestion
                driver = picking.move_ids.mapped("move_dest_ids.picking_id.driver_id")
```

**Review by Shide — APPROVED** [review]
Code LGTM. 

**question:** mapped should contain `move_dest_ids` (o2m > o2m)?

File: delivery_driver/migrations/16.0.1.0.1/post-migration.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+from odoo import SUPERUSER_ID, api
+
+_logger = logging.getLogger(__name__)
+
+
+def migrate(cr, version):
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        partners = env["carrier.driver"].search([("driver_id", "!=", False)])
```
Comment (yajo) [inline delivery_driver/migrations/16.0.1.0.1/post-migration.py]: Don't you want partners?

```suggestion
        partners = env["carrier.driver"].search([("driver_id", "!=", False)]).mapped("driver_id")
```

File: delivery_driver/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            [("driver_id", "in", self.ids)]
+        )
+        if drivers_in_carrier and "is_driver" in vals and not vals.get("is_driver"):
+            raise ValidationError(
+                _(
+                    "You can't remove the driver flag from a partner that"
+                    " is set as driver in a delivery method."
+                )
+            )
+        return super().write(vals)
```
Comment (yajo) [inline delivery_driver/models/res_partner.py]: issue: This looks more like a constraint. Better to implement it as such.

File: delivery_driver/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    def _get_name(self):
+        if self.env.context.get("show_driver"):
```
Comment (yajo) [inline delivery_driver/models/res_partner.py:30]: suggestion: Please add a little docstring about why this is needed.

File: delivery_driver/models/stock_picking.py [hunk 5]
```diff
@@ -11,14 +11,19 @@ class StockPicking(models.Model):
...
+    @api.depends("carrier_id", "move_ids.move_dest_ids.picking_id.driver_id")
     def _compute_driver_id(self):
         for picking in self:
             if picking.state not in {"done", "cancel"}:
-                picking.driver_id = picking.carrier_id.driver_id
+                driver = picking.move_ids.mapped("move_dest_ids.picking_id.driver_id")
+                if driver:
+                    picking.driver_id = driver[0]
+                else:
+                    picking.driver_id = picking.carrier_id.driver_id
```
Comment (yajo) [inline delivery_driver/models/stock_picking.py]: suggestion: you can use a slice to pick the 1st record or none, as a shortcut:

```suggestion
                picking.driver_id = driver[:1] or picking.carrier_id.driver_id
```

**Review by yajo — CHANGES_REQUESTED** [review]
Some blocking issues.

File: delivery_driver/migrations/16.0.1.0.1/post-migration.py (hunk 2 above)
Comment (EmilioPascual) [reply delivery_driver/migrations/16.0.1.0.1/post-migration.py]: Ouch!! I updated this. Thanks!

**Review by yajo — APPROVED** [review]

**Comment by EmilioPascual** [comment]
@rafaelbn please review and merge if it is ok

**Review by rafaelbn — APPROVED** [review]
2 weeks PR 1 week in prod and tested un staging!

**Pull request merged**